### PR TITLE
"What's New" credit and referencing

### PR DESCRIPTION
**Pull request by trexfeathers**
Adds two new aspects to Iris "What's New" entries to improve Iris' efficacy as an open source project.
* **Pull Request references**
Open source projects need to be both dynamic and transparent, so it's important that anyone interested can get as much detail as they need on changes in the project.
* **Crediting developers**
Open source communities rely on their developer base, some of whom might be volunteering their time. Public recognition rewards contributing developers and demonstrates the quality of the community, hopefully encouraging further and wider collaboration.

Mimicking the [Cartopy format](https://scitools.org.uk/cartopy/docs/latest/whats_new.html) - developer credit first - meant adapting the 'voice' of many of the existing "What's New" entries to get them to state the implemented change before any further context. If anyone below feels I have minced their words, **or if they are unhappy being publicly credited**, feel free to speak up!
* @MoseleyS 
* @stephenworsley 
* @pp-mo 
* @abooton 
* @bouweandela 
* @bjlittle 
* @trexfeathers 
* @jonseddon 
* @tkknight

### CODE REVIEW

**Comment by trexfeathers** [comment]
Closes #3818

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (bouweandela) [inline docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:31]: An alternative way of doing this could be to use the names contributors use publicly on GitHub (and link to their profile page), as done e.g. here: https://docs.dask.org/en/latest/changelog.html#id16. That has the advantage that it's up to contributors themselves to decide if they want to be listed with their real name.

Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:31]: Thanks @bouweandela , I had considered this but I only checked Cartopy, SciPy and Numpy so I hadn't seen any precedent. Good to know that Dask do it - I will make the change

Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst]: 596f31b7aa678cecb4903b0898cbbeef73aba594

**Comment by trexfeathers** [comment]
@bouweandela as one of our newest core devs do you fancy reviewing and merging your first Iris PR?

**Comment by trexfeathers** [comment]
Ported over the contents of #3843 - felt like I was asking for Git trouble otherwise!

Closes #3837 

Adds:
* More What's New entries for important pre- 4761a10630b044aedb38f56f96ecff3f52085ca4 work that had been missed
  Contributions from @stephenworsley , @trexfeathers , @bjlittle , @pp-mo , @lbdreyer (in addition to credits in #3838 )

---

* Consistent "What's New" ordering and headings between docs, template and `latest.rst`
  With _Internal_ now at the bottom
* Updated docs on what is expected regarding "What's New" entries
* Updated docs pull request checklist to bring in line with current verbally agreed practices
* Added pull request template referencing the docs pull request checklist

File: .github/pull_request_template.md
Comment (bjlittle) [inline .github/pull_request_template.md:5]: @trexfeathers Nice! I was thinking along the lines of something like this...


## 🚀 Pull Request

### Description





### Additional information

<details>
<summary>Click to expand <b>this section...</b></summary>

```
Add additional verbose information in a collapsible section.
```
See [here](https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab) for further details.
</details>

---
[Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/pulls.html#the-iris-check-list)

Comment (bjlittle) [reply .github/pull_request_template.md:5]: What do you think?

Comment (trexfeathers) [reply .github/pull_request_template.md:5]: I'm not sure what problems this solves. We don't have PR categories so the title is purely aesthetic, and it's pretty rare for a PR to be so detailed that it needs headings/collapsing. Indeed those 5% come from devs who probably know these features already. 

I'm worried we're just going to encourage a load of markdown 'artifacts' on the 95% of the PR's that are small enough to be better without them. Are you keen for/expecting PR's to get more detailed in future? 

Now I've seen it I quite like the idea of _only_ including the checklist link in the way you have written it.

Comment (bjlittle) [reply .github/pull_request_template.md:5]: I personally find this pretty minimal myself. If anything you could drop the `Additional information` section... I think you're right, it would most likely be "just deleted" by most PR authors.

However, what I am keen that we do is have is a consistent branding on GitHub for `iris`, and as such this template suggestion is pretty much in line with the new [Issue templates](https://github.com/SciTools/iris/issues/new/choose).

So, we could trim it down to simply...

## 🚀 Pull Request

### Description




---
[Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/pulls.html#the-iris-check-list)

Comment (bjlittle) [reply .github/pull_request_template.md:5]: BTW I think that the actual [Iris pull request check list](https://scitools-iris.readthedocs.io/en/latest/developers_guide/pulls.html#the-iris-check-list) needs a bit of work... but @tkknight may have that in hand :thinking:

Comment (trexfeathers) [reply .github/pull_request_template.md:5]: If consistency is the key factor then I appreciate where you're coming from. I'll make the change either today or Monday

> BTW I think that the actual Iris pull request check list needs a bit of work...

Is that even after [the changes to the checklist that this PR suggests](https://github.com/SciTools/iris/pull/3838/commits/15ebb9965d6552900b7fd881bfcb3b729681dd5e#diff-8cd7796992d8299a5dbab9c6080157eb)?

Comment (bjlittle) [reply .github/pull_request_template.md:5]: Oops, sorry... Hasn't got that far 🤭

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (rcomer) [inline docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:31]: Sometimes the review can take a while, e.g. because no reviewers are available or because the change turns out to be more complicated than the developer realised (the latter happens to me a lot!)  I wonder if it would be better to leave the whatsnew PR until the main one is almost done.

Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:31]: I agree this is a problem. But it would be ideal if the reviewer would still be able to insist on a "What's New" before merging. Do you think it would work for the timeline to look something like: "OK I'm now happy we've sorted the code changes, please can you create a "What's New" PR and we'll be done."?

Comment (rcomer) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:31]: That seems reasonable to me.  Having said that, if the main PR is almost ready to merge, it might be simpler to rebase and add the whatsnew to the existing branch.

Rebase would also ensure that Travis is looking at the right thing.  I did once manage to [break master with a failure that didn’t show up in my PR](https://github.com/SciTools/iris/pull/2960#issuecomment-385633071). 😬

File: .github/pull_request_template.md
Comment (tkknight) [reply .github/pull_request_template.md:5]: FYI, the **developers guide** overhaul now has a branch in my forked iris repo.  It still needs some work.  The PR checklist has been improved I can **rebase** if it is updated in @trexfeathers pr.

- [My branch](https://github.com/SciTools/iris/compare/master...tkknight:developer_guide_overhaul)
- [My branch pr checklist](https://github.com/SciTools/iris/compare/master...tkknight:developer_guide_overhaul#diff-107170be3096da8087bc84d4a3018b83)

File: docs/iris/src/whatsnew/latest.rst
Comment (rcomer) [inline docs/iris/src/whatsnew/latest.rst:72]: I haven’t checked, but it looks like this fixed #3347.

Comment (rcomer) [reply docs/iris/src/whatsnew/latest.rst:72]: I've now checked with `master` and #3347 is indeed fixed :sunglasses:.

Comment (trexfeathers) [reply docs/iris/src/whatsnew/latest.rst:72]: Wow, the "What's New" is working even before it gets merged 😂

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:31]: >Rebase would also ensure that Travis is looking at the right thing. I did once manage to break master with a failure that didn’t show up in my PR. 😬

As I understand, Travis simulates a merge with the default branch, then runs the tests. So rather than rebasing I **think** the problem you reference would have been caught by re-spinning Travis just before merging the PR.

File: docs/iris/src/developers_guide/pulls.rst
Comment (trexfeathers) [inline docs/iris/src/developers_guide/pulls.rst:1]: Changes proposed to this file are the minimum to bring the check list up to date. I expect them to be soon over-written by @tkknight 's wholesale re-write of the Developer's Guide - [see this comment](https://github.com/SciTools/iris/pull/3838#discussion_r487403532)

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:28]: 8158666c0a8432af262e9141c7167975501a61b4

File: .github/pull_request_template.md
Comment (trexfeathers) [reply .github/pull_request_template.md:8]: @bjlittle 08ee8ba8fb32244e7c566138dfae0b7beb4f801d

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (stephenworsley) [inline docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:95]: There should be a comma after "references".

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (stephenworsley) [inline docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:133]: There should be a full stop after "etc".

File: .github/pull_request_template.md
Comment (bouweandela) [reply .github/pull_request_template.md:8]: Maybe "Description" could be replaced by "What's new" and the content of that section could then be used to compose the complete "What's new" file when making a new release by combining all the entries from merged pull requests since the previous release, as also proposed here https://github.com/SciTools/iris/issues/3837#issuecomment-691956633? To keep it easy to use, harvesting entries and combining them into a latest.rst file would then be done by a script using the GitHub API.

File: docs/iris/src/whatsnew/latest.rst
Comment (stephenworsley) [inline docs/iris/src/whatsnew/latest.rst:42]: The comma here should be removed.

File: docs/iris/src/whatsnew/latest.rst
Comment (stephenworsley) [inline docs/iris/src/whatsnew/latest.rst:98]: There should be a space between "and" and ":meth:"

File: docs/iris/src/whatsnew/latest.rst
Comment (stephenworsley) [inline docs/iris/src/whatsnew/latest.rst:132]: "is be an" should be "to be an"

File: docs/iris/src/whatsnew/latest.rst
Comment (stephenworsley) [inline docs/iris/src/whatsnew/latest.rst:146]: Comma should probably be removed

**Review by stephenworsley — CHANGES_REQUESTED** [review]
Mostly just a couple typos to fix. Otherwise this is looking good. Good work!

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:42]: Good point @stephenworsley . This workflow is an effort to avoid needing to rebase, and there's every chance that a rebase to the "What's New" PR might be followed by actual "What's New" corrections and it would all get pushed up.

I don't see an easy solution, and in the long term I'm personally in favour of @bouweandela 's proposal (https://github.com/SciTools/iris/issues/3837#issuecomment-691956633), so in the meantime do you think it would be OK to just make specific reference to Travis' linkcheck test as being the thing that `should now be working`?

Comment (stephenworsley) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:42]: That sounds fair to me.

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst]: 3138a01bbee3e9e2c67644ff9b4eae4eddde73aa

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst]: 3138a01bbee3e9e2c67644ff9b4eae4eddde73aa

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst]: 3138a01bbee3e9e2c67644ff9b4eae4eddde73aa

File: docs/iris/src/whatsnew/latest.rst
Comment (trexfeathers) [reply docs/iris/src/whatsnew/latest.rst]: d1038c403d23de9e461e3fe5e3a94cb7def73778

File: docs/iris/src/whatsnew/latest.rst
Comment (trexfeathers) [reply docs/iris/src/whatsnew/latest.rst]: d1038c403d23de9e461e3fe5e3a94cb7def73778

File: docs/iris/src/whatsnew/latest.rst
Comment (trexfeathers) [reply docs/iris/src/whatsnew/latest.rst]: d1038c403d23de9e461e3fe5e3a94cb7def73778

File: docs/iris/src/whatsnew/latest.rst
Comment (trexfeathers) [reply docs/iris/src/whatsnew/latest.rst]: d1038c403d23de9e461e3fe5e3a94cb7def73778

File: docs/iris/src/developers_guide/documenting/whats_new_contributions.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/whats_new_contributions.rst]: 544d324c4cc638fbccc06cbcde9e05a170a1a205

File: .github/pull_request_template.md
Comment (trexfeathers) [reply .github/pull_request_template.md:8]: Totally 👍 on this (https://github.com/SciTools/iris/issues/3837#issuecomment-691980738), but I'm very reluctant to introduce something new to the docs build at such a late stage for v3.0.0

**Review by stephenworsley — APPROVED** [review]

File: .github/pull_request_template.md
Comment (bouweandela) [reply .github/pull_request_template.md:8]: It's up to you, you know your own timeline best.

**Comment by trexfeathers** [comment]
Will need a follow up PR to correct the ["What's New" from #3804 ](https://github.com/SciTools/iris/pull/3804/files#diff-7f35cedb5073b58582e2a3fa56496ca2R95-R99)

**Comment by pp-mo** [comment]
I see that you are mostly using full real names, but not for everyone, e.g [here](https://github.com/SciTools/iris/pull/3838/files#diff-7f35cedb5073b58582e2a3fa56496ca2R266) and [here](https://github.com/SciTools/iris/pull/3838/files#diff-7f35cedb5073b58582e2a3fa56496ca2R217).
I assume you have based this on the public user account details, hence e.g. @lbdreyer is not identified.  
I guess that is ok + means you don't have to ask everyone what they are happy with.

However. I think it might look better, and be safer in terms of what contributors are really comfortable with, to **just use user ids**.  
Certainly, I always stick to userids in discussions, where I don't want to give any appearance of relying on offline relations.

**Comment by trexfeathers** [comment]
> I see that you are mostly using full real names, but not for everyone, e.g here and here.
I assume you have based this on the public user account details, hence e.g. @lbdreyer is not identified.
I guess that is ok + means you don't have to ask everyone what they are happy with.

https://github.com/SciTools/iris/pull/3838#discussion_r486827236

**Pull request merged**
